Applications/Spire/Source/Ui/TagBox.cpp

```diff
@@ -0,0 +1,480 @@
+using namespace Spire::Styles;
+
+namespace {
+  auto INPUT_BOX_STYLE(StyleSheet style) {
+    style.get(Any()).
+      set(vertical_padding(scale_height(3)));
+    return style;
+  }
+
+  auto LIST_ITEM_STYLE() {
+    auto style = StyleSheet();
+    style.get(Any()).
+      set(BackgroundColor(QColor(Qt::transparent))).
+      set(border_size(0)).
+      set(padding(0));
+    return style;
+  }
+
+  auto LIST_VIEW_STYLE(StyleSheet style) {
+    style.get(Any()).
+      set(ListItemGap(scale_width(4))).
+      set(ListOverflowGap(scale_width(3))).
+      set(Overflow::WRAP).
+      set(Qt::Horizontal);
+    return style;
+  }
+
+  auto TEXT_BOX_STYLE(StyleSheet style) {
+    style.get(Any() || Hover() || Focus() || Disabled()).
+      set(BackgroundColor(QColor(Qt::transparent))).
+      set(border_size(0)).
+      set(horizontal_padding(0)).
+      set(vertical_padding(scale_height(2)));
+    return style;
+  }
+}
+
+struct TagBox::PartialListModel : public ListModel {
+  std::shared_ptr<ListModel> m_source;
+
+  explicit PartialListModel(std::shared_ptr<ListModel> source)
+    : m_source(std::move(source)) {}
+
+  virtual int get_size() const {
+    return m_source->get_size() + 2;
+  }
+
+  virtual const std::any& at(int index) const {
+    if(index < 0 || index >= get_size()) {
+      throw std::out_of_range("The index is out of range.");
+    }
+    if(index < m_source->get_size()) {
+      return m_source->at(index);
+    }
+    static auto value = std::any();
+    return value;
+  }
+
+  virtual QValidator::State set(int index, const std::any& value) {
+    if(index < 0 || index >= get_size()) {
+      return QValidator::State::Invalid;
+    }
+    if(index < m_source->get_size()) {
+      return m_source->set(index, value);
+    }
+    return QValidator::State::Acceptable;
+  }
+
+  connection connect_operation_signal(
+    const OperationSignal::slot_type& slot) const override {
+    return m_source->connect_operation_signal(slot);
+  }
+};
+
+TagBox::TagBox(std::shared_ptr<ListModel> list_model,
+    std::shared_ptr<TextModel> current_model, QWidget* parent)
+    : QWidget(parent),
+      m_model(std::make_shared<PartialListModel>(std::move(list_model))),
+      m_focus_observer(*this),
+      m_overflow(TagBoxOverflow::WRAP),
+      m_tags_width(0),
+      m_list_item_gap(0) {
+  auto layout = new QHBoxLayout(this);
+  layout->setContentsMargins({});
+  m_text_box = new TextBox(std::move(current_model));
+  set_style(*m_text_box, TEXT_BOX_STYLE(get_style(*m_text_box)));
+  m_text_box->installEventFilter(this);
+  m_list_view = new ListView(m_model,
+    std::bind_front(&TagBox::build_tag, this));
+  m_list_view->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+  set_style(*m_list_view, LIST_VIEW_STYLE(get_style(*m_list_view)));
+  m_list_view_style_connection = connect_style_signal(*m_list_view,
+    std::bind_front(&TagBox::on_list_view_style, this));
+  for(auto i = 0; i < m_list_view->get_list()->get_size(); ++i) {
```

Comment:
Setting style like this is very fragile. Use a combinator instead, refer to CSS for how to style children:

https://developer.mozilla.org/en-US/docs/Web/CSS/Child_combinator